spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java

```diff
@@ -156,6 +158,16 @@ public void close() {
 		catch (IOException ex) {
 			throw new UncheckedIOException("failed to close an SFTP client", ex);
 		}
+
+		try {
+			ClientSession session = this.sftpClient.getSession();
+			if (session != null && session.isOpen()) {
```

Comment:
I don't think this fix is correct.
The `DefaultSftpClient` is based on the externally provided `ClientSession`.
We cannot mutate that one from our volatile `SftpSession`.

Let me double check what is really the problem!

Reply:
OK. I confirm the problem, but sounds more like we have to fix `DefaultSftpSessionFactory` to keep cache for the `ClientSession` and reuse it for every `ConcurrentSftpClient` created.
If that is closed, we just have renew it.
Then we have to implement `DisposableBean` over there to close this shared `ClientSession` properly.
And apparently we also have to `SshClient.stop()` if it `isInnerClient`.